text/2018-01-14-skiplist.md

```diff
@@ -0,0 +1,257 @@
+* [`ConcurrentSkipListMap<long, long> (2 threads)`](https://gist.github.com/stjepang/74d1abc7230ad6e6dd0c4aec1f4cab4b): 1353 ms
+
+The bottom line is: in single-threaded scenarios `SkipMap` should be comparable in performance
+to any typical binary search tree (although not to a B-tree). As we add more threads, it seems to
+scale quite well. I don't have a machine with a high number of cores to test scalability more thoroughly,
+but these numbers seem promising so far nonetheless.
+
+### Iteration
+
+The skip list supports easy iteration. Note that when iterating over a `SkipMap` we hand out an `Entry`
+for each entry in it. Creating an entry involves incrementing its reference count, and when moving
+from one entry to another we also have to pin the current thread. This is a lot of reference count
+updating and a lot of pinning.
+
+Here are some benchmark numbers for iterating over a million randomly inserted entries:
+
+* `BTreeMap` (Rust): 18 ms
+* `SkipMap` (Rust): 113 ms
+* `std::map` (C++): 93 ms
+* `TreeMap` (Java): 41 ms
+* `ConcurrentSkipListMap` (Java): 32 ms
+
+Interesting observations:
+
+* Iteration over a `BTreeMap` is very fast - this shouldn't be surprising since adjacent
+  elements are grouped into blocks.
+
+* `SkipMap` is the slowest map here. I've tried measuring how long iteration takes withut
+  reference count updating and without pinning, and it turns out to be around 95 ms. That is
+  very similar to `std::map` in C++. Also, reference counting and pinning definitely incurs
+  a measurable cost, but it's not a *terrible* one.
+
+* Java is fast - even `TreeMap` is faster than `std::map` in C++. How is that possible?
+  Well, the answer lies in the fact that Java's GC kicks from time to time, moves
+  allocated nodes in memory (it's a compacting GC), and tries to lay out linked nodes
+  as close as possible, thus optimizing for cache efficiency.
+
+* Let's try tuning Java's GC by using option `-XX:NewSize=1024m`. This option sets the
+  size of the new generation to 1024 MB (a huge number), which means compaction should
+  never kick in. Indeed, iteration timings are much different now - with `TreeMap` it takes
+  124 ms and with `ConcurrentSkipListMap` it takes 110 ms. Now that's much closer to `SkipMap`
+  and `std::map`.
+
+# Drawbacks
+
+Skip lists are not very exciting when it comes to performance. Hash tables, B-trees (Bw-Tree is
+a lock-free B-tree variant), and radix trees (ART - adaptive radix tree can be made concurrent)
+are usually more performant. However, these faster data structures are not as general as skip lists
+and have to make sacrifices by restricting the set of supported operations or by making the API less ergonomic.
+
+# Alternatives
+
+A few possible alternatives to the lock-free skip list might be:
+
+1. Adaptive radix tree (keys can only be byte arrays).
+2. Skip tree (moves elements in memory, thus constraining the API).
+2. Bw-Tree (moves elements in memory, thus constraining the API).
```

Comment:
Can I ask why skip tree and Bw-tree have contrained API?  Even elements are moved, I think we can support a similar API to the one presented here.

Reply:
Sorry, this question somehow fell off my radar. Consider this:

```rust
struct Key(String, AtomicUsize);

// Also impl PartialOrd, PartialEq, and Eq...
impl Ord for Key {
    fn ord(&self, other: &Key) -> Ordering { self.0.cmp(&other.0) }
}

let set = BwTreeSet::<Key>::new();

crossbeam::scope(|s| {
    s.spawn(|| {
        loop {
            for e in set.iter() {
                e.1.fetch_add(1, Relaxed);
            }
        }
    });

    s.spawn(|| {
        loop {
            let s = random_string();
            // Some elements might be moved into new blocks during insertion.
            set.insert(Key(s, AtomicUsize::new(0)));
        }
    });
});
```

What should happen if one thread is incrementing an `AtomicUsize`, while another thread is moving it to a new location? That's not going to work, so we'll have to impose some kind of restriction on the API.

For example, we could say that all keys must be `Copy`, thus forbidding types like `Key`. Or, another solution might be to implement iterators so that they return clones of elements rather than references to elements - that way threads couldn't simultaneously move and modify an element.